qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -104,24 +104,54 @@ class OptimizationResult:
         results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
+        variables: The list of variables under optimization.
     """
 
     Status = OptimizationResultStatus
 
     def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
-        self._val = x
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
+                 variables: Optional[List[Variable]] = None) -> None:
+        self._x = x  # pylint: disable=invalid-name
+        self._variables = variables
         self._fval = fval
         self._results = results
         self._status = status
 
     def __repr__(self):
-        return '([{}] / {} / {})'.format(','.join([str(x_) for x_ in self.x]), self.fval,
-                                         self.status)
+        return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
+            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
 
-    def __str__(self):
-        return 'x=[{}], fval={}'.format(','.join([str(x_) for x_ in self.x]), self.fval)
+    def __getitem__(self, item: Union[int, str]):
+        if isinstance(item, int):
+            return self.x[item]
+        if isinstance(item, str):
+            return self.variables[item]
+        raise QiskitOptimizationError("Integer or string parameter required, instead "
+                                      + type(item).__name__ + " provided.")
+
+    @property
+    def variables(self) -> Optional[Dict[str, int]]:
+        """Returns the pairs of variable names and values under optimization.
```

Comment:
This is very confusing with `QuadraticProgram` where the same property returns a list of variables, not a dictionary. I'd suggest to have a similar signature.

Reply:
What do you suggest for a method name for a dictionary of variable names and values? I'm wondering a good name.

Reply:
I'd rather see `def variables(self) -> List[Variable]` since such a list is passed in the constructor anyway.

Reply:
Then, how about removing a method to return a dict and rename `x_names` as `variables`?

Reply:
You mean keeping `variables` as `List[Variable]`? Makes sense.